File: Leetcode/238. Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int i=nums.length;
        int z=0;
        int product=1,temp=1;
        int[] result=new int[nums.length]; 
        for(int n:nums){
            product*=n;
        }
        if(product==0){
            for(int n:nums){
                if(n==0) {
                    z++;
                    continue;
                }
                temp*=n;
            }
        }
        for(int j=0;j<nums.length;j++){
            if(z>=2){
                result[j]=0;
            }
            if(z==1){
                if(nums[j]==0){
                  result[j]=temp;
                }
                else{
                    result[j]=0;
                }
            }
            if(z==0){
                result[j]=product/nums[j];
            }
        }
        return result;
    }
}

File: C /238. Product of Array Except Self.c


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* productExceptSelf(int* nums, int numsSize, int* returnSize){
    int i;
    *returnSize = numsSize;
    int *result = (int *)malloc(sizeof(int)*numsSize);
    result[0] = 1;
    for(i=1; i<numsSize; i++) {
        result[i] = nums[i-1] * result[i-1];
    }
    int t = 1;
    for(i=numsSize-1; i>=0; i--) {
        result[i] *= t;
        t *= nums[i];
    }
    return result;
}
